attest/attest_monitor.go

```diff
@@ -0,0 +1,82 @@
+package attest
+
+import (
+	"sync"
+	"time"
+
+	"github.com/bnb-chain/greenfield-challenger/db/model"
+	"github.com/bnb-chain/greenfield-challenger/executor"
+	"github.com/bnb-chain/greenfield-challenger/logging"
+)
+
+type AttestMonitor struct {
+	executor             *executor.Executor
+	mtx                  sync.RWMutex
+	attestedChallengeIds map[uint64]bool // used to save the last attested challenge id
+	dataProvider         DataProvider
+}
+
+func NewAttestMonitor(executor *executor.Executor, dataProvider DataProvider) *AttestMonitor {
+	return &AttestMonitor{
+		executor:             executor,
+		mtx:                  sync.RWMutex{},
+		attestedChallengeIds: make(map[uint64]bool, 0),
+		dataProvider:         dataProvider,
+	}
+}
+
+// UpdateAttestedChallengeIdLoop polls the blockchain for latest attested challengeIds and updates their status
+func (a *AttestMonitor) UpdateAttestedChallengeIdLoop() {
+	ticker := time.NewTicker(QueryAttestedChallengeInterval)
+	queryCount := 0
+	for range ticker.C {
+		challengeIds, err := a.executor.QueryLatestAttestedChallengeIds()
+		// logging.Logger.Infof("latest attested challenge ids: %+v", challengeIds)
+		if err != nil {
+			logging.Logger.Errorf("update latest attested challenge error, err=%+v", err)
+			continue
+		}
+		logging.Logger.Infof("latest attested challenge ids: %+v", challengeIds)
+		a.mtx.Lock()
+		a.updateAttestedCacheAndEventStatus(a.attestedChallengeIds, challengeIds)
```

Comment:
In each call, it will return all the 300 latest attested challenges. We should use some cache to filter out the ids which had been written to database. Otherwise, there are lots of useless database write.

Reply:
fixed